crates/net/common/src/ratelimit.rs

```diff
@@ -0,0 +1,151 @@
+//! A rate limit implementation to enforce a specific rate.
+
+use std::{
+    future::Future,
+    pin::Pin,
+    task::{Context, Poll},
+    time::Duration,
+};
+use tokio::time::Sleep;
+
+/// Given a [Rate] this type enforces a rate limit.
+#[derive(Debug)]
+pub struct RateLimit {
+    rate: Rate,
+    state: State,
+    sleep: Pin<Box<Sleep>>,
+}
+
+// === impl RateLimit ===
+
+impl RateLimit {
+    /// Create a new rate limiter
+    pub fn new(rate: Rate) -> Self {
+        let until = tokio::time::Instant::now();
+        let state = State::Ready { until, remaining: rate.limit() };
+
+        RateLimit { rate, state, sleep: Box::pin(tokio::time::sleep_until(until)) }
+    }
+
+    /// Checks if the [RateLimit] is ready to handle a new call
+    pub fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<()> {
```

Comment:
Is the intention that you call `poll_ready(cx)` every time before calling `tick`? I.e. let's say I wanted to send a request, I would need to do:

```
if rate_limiter.poll_ready(cx).is_ready() {
  // send req
  rate_limiter.tick();
}
```

And this is only usable in a hand-rolled future, right? Should we not have an API for things that do not have explicit access to `cx`?

Reply:
yes,

only for future::poll, but can use `poll_fn` though, unsure about async code yet